// @protobuf-ts/plugin-framework/build/es2015/google/protobuf/descriptor.js
// @generated by protobuf-ts 1.0.0-alpha.35 with parameters force_optimize_code_size,long_type_string
// @generated from protobuf file "google/protobuf/descriptor.proto" (package "google.protobuf", syntax proto2)
// tslint:disable
//
// Protocol Buffers - Google's data interchange format
// Copyright 2008 Google Inc.  All rights reserved.
// https://developers.google.com/protocol-buffers/
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//
// Author: kenton@google.com (Kenton Varda)
//  Based on original Protocol Buffers design by
//  Sanjay Ghemawat, Jeff Dean, and others.
//
// The messages in this file describe the definitions found in .proto files.
// A valid .proto file can be translated directly to a FileDescriptorProto
// without any other information (e.g. without reading its imports).
//
import { ScalarType } from "@protobuf-ts/runtime";
import { RepeatType } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
/**
 * @generated from protobuf enum google.protobuf.FieldDescriptorProto.Type
 */
export var FieldDescriptorProto_Type;
(function (FieldDescriptorProto_Type) {
    /**
     * @generated synthetic value - protobuf-ts requires all enums to have a 0 value
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["UNSPECIFIED$"] = 0] = "UNSPECIFIED$";
    /**
     * 0 is reserved for errors.
     * Order is weird for historical reasons.
     *
     * @generated from protobuf enum value: TYPE_DOUBLE = 1;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["DOUBLE"] = 1] = "DOUBLE";
    /**
     * @generated from protobuf enum value: TYPE_FLOAT = 2;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["FLOAT"] = 2] = "FLOAT";
    /**
     * Not ZigZag encoded.  Negative numbers take 10 bytes.  Use TYPE_SINT64 if
     * negative values are likely.
     *
     * @generated from protobuf enum value: TYPE_INT64 = 3;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["INT64"] = 3] = "INT64";
    /**
     * @generated from protobuf enum value: TYPE_UINT64 = 4;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["UINT64"] = 4] = "UINT64";
    /**
     * Not ZigZag encoded.  Negative numbers take 10 bytes.  Use TYPE_SINT32 if
     * negative values are likely.
     *
     * @generated from protobuf enum value: TYPE_INT32 = 5;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["INT32"] = 5] = "INT32";
    /**
     * @generated from protobuf enum value: TYPE_FIXED64 = 6;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["FIXED64"] = 6] = "FIXED64";
    /**
     * @generated from protobuf enum value: TYPE_FIXED32 = 7;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["FIXED32"] = 7] = "FIXED32";
    /**
     * @generated from protobuf enum value: TYPE_BOOL = 8;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["BOOL"] = 8] = "BOOL";
    /**
     * @generated from protobuf enum value: TYPE_STRING = 9;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["STRING"] = 9] = "STRING";
    /**
     * Tag-delimited aggregate.
     * Group type is deprecated and not supported in proto3. However, Proto3
     * implementations should still be able to parse the group wire format and
     * treat group fields as unknown fields.
     *
     * @generated from protobuf enum value: TYPE_GROUP = 10;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["GROUP"] = 10] = "GROUP";
    /**
     * Length-delimited aggregate.
     *
     * @generated from protobuf enum value: TYPE_MESSAGE = 11;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["MESSAGE"] = 11] = "MESSAGE";
    /**
     * New in version 2.
     *
     * @generated from protobuf enum value: TYPE_BYTES = 12;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["BYTES"] = 12] = "BYTES";
    /**
     * @generated from protobuf enum value: TYPE_UINT32 = 13;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["UINT32"] = 13] = "UINT32";
    /**
     * @generated from protobuf enum value: TYPE_ENUM = 14;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["ENUM"] = 14] = "ENUM";
    /**
     * @generated from protobuf enum value: TYPE_SFIXED32 = 15;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["SFIXED32"] = 15] = "SFIXED32";
    /**
     * @generated from protobuf enum value: TYPE_SFIXED64 = 16;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["SFIXED64"] = 16] = "SFIXED64";
    /**
     * Uses ZigZag encoding.
     *
     * @generated from protobuf enum value: TYPE_SINT32 = 17;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["SINT32"] = 17] = "SINT32";
    /**
     * Uses ZigZag encoding.
     *
     * @generated from protobuf enum value: TYPE_SINT64 = 18;
     */
    FieldDescriptorProto_Type[FieldDescriptorProto_Type["SINT64"] = 18] = "SINT64";
})(FieldDescriptorProto_Type || (FieldDescriptorProto_Type = {}));
/**
 * @generated from protobuf enum google.protobuf.FieldDescriptorProto.Label
 */
export var FieldDescriptorProto_Label;
(function (FieldDescriptorProto_Label) {
    /**
     * @generated synthetic value - protobuf-ts requires all enums to have a 0 value
     */
    FieldDescriptorProto_Label[FieldDescriptorProto_Label["UNSPECIFIED$"] = 0] = "UNSPECIFIED$";
    /**
     * 0 is reserved for errors
     *
     * @generated from protobuf enum value: LABEL_OPTIONAL = 1;
     */
    FieldDescriptorProto_Label[FieldDescriptorProto_Label["OPTIONAL"] = 1] = "OPTIONAL";
    /**
     * @generated from protobuf enum value: LABEL_REQUIRED = 2;
     */
    FieldDescriptorProto_Label[FieldDescriptorProto_Label["REQUIRED"] = 2] = "REQUIRED";
    /**
     * @generated from protobuf enum value: LABEL_REPEATED = 3;
     */
    FieldDescriptorProto_Label[FieldDescriptorProto_Label["REPEATED"] = 3] = "REPEATED";
})(FieldDescriptorProto_Label || (FieldDescriptorProto_Label = {}));
/**
 * Generated classes can be optimized for speed or code size.
 *
 * @generated from protobuf enum google.protobuf.FileOptions.OptimizeMode
 */
export var FileOptions_OptimizeMode;
(function (FileOptions_OptimizeMode) {
    /**
     * @generated synthetic value - protobuf-ts requires all enums to have a 0 value
     */
    FileOptions_OptimizeMode[FileOptions_OptimizeMode["UNSPECIFIED$"] = 0] = "UNSPECIFIED$";
    /**
     * Generate complete code for parsing, serialization,
     *
     * @generated from protobuf enum value: SPEED = 1;
     */
    FileOptions_OptimizeMode[FileOptions_OptimizeMode["SPEED"] = 1] = "SPEED";
    /**
     * etc.
     *
     * Use ReflectionOps to implement these methods.
     *
     * @generated from protobuf enum value: CODE_SIZE = 2;
     */
    FileOptions_OptimizeMode[FileOptions_OptimizeMode["CODE_SIZE"] = 2] = "CODE_SIZE";
    /**
     * Generate code using MessageLite and the lite runtime.
     *
     * @generated from protobuf enum value: LITE_RUNTIME = 3;
     */
    FileOptions_OptimizeMode[FileOptions_OptimizeMode["LITE_RUNTIME"] = 3] = "LITE_RUNTIME";
})(FileOptions_OptimizeMode || (FileOptions_OptimizeMode = {}));
/**
 * @generated from protobuf enum google.protobuf.FieldOptions.CType
 */
export var FieldOptions_CType;
(function (FieldOptions_CType) {
    /**
     * Default mode.
     *
     * @generated from protobuf enum value: STRING = 0;
     */
    FieldOptions_CType[FieldOptions_CType["STRING"] = 0] = "STRING";
    /**
     * @generated from protobuf enum value: CORD = 1;
     */
    FieldOptions_CType[FieldOptions_CType["CORD"] = 1] = "CORD";
    /**
     * @generated from protobuf enum value: STRING_PIECE = 2;
     */
    FieldOptions_CType[FieldOptions_CType["STRING_PIECE"] = 2] = "STRING_PIECE";
})(FieldOptions_CType || (FieldOptions_CType = {}));
/**
 * @generated from protobuf enum google.protobuf.FieldOptions.JSType
 */
export var FieldOptions_JSType;
(function (FieldOptions_JSType) {
    /**
     * Use the default type.
     *
     * @generated from protobuf enum value: JS_NORMAL = 0;
     */
    FieldOptions_JSType[FieldOptions_JSType["JS_NORMAL"] = 0] = "JS_NORMAL";
    /**
     * Use JavaScript strings.
     *
     * @generated from protobuf enum value: JS_STRING = 1;
     */
    FieldOptions_JSType[FieldOptions_JSType["JS_STRING"] = 1] = "JS_STRING";
    /**
     * Use JavaScript numbers.
     *
     * @generated from protobuf enum value: JS_NUMBER = 2;
     */
    FieldOptions_JSType[FieldOptions_JSType["JS_NUMBER"] = 2] = "JS_NUMBER";
})(FieldOptions_JSType || (FieldOptions_JSType = {}));
/**
 * Is this method side-effect-free (or safe in HTTP parlance), or idempotent,
 * or neither? HTTP based RPC implementation may choose GET verb for safe
 * methods, and PUT verb for idempotent methods instead of the default POST.
 *
 * @generated from protobuf enum google.protobuf.MethodOptions.IdempotencyLevel
 */
export var MethodOptions_IdempotencyLevel;
(function (MethodOptions_IdempotencyLevel) {
    /**
     * @generated from protobuf enum value: IDEMPOTENCY_UNKNOWN = 0;
     */
    MethodOptions_IdempotencyLevel[MethodOptions_IdempotencyLevel["IDEMPOTENCY_UNKNOWN"] = 0] = "IDEMPOTENCY_UNKNOWN";
    /**
     * implies idempotent
     *
     * @generated from protobuf enum value: NO_SIDE_EFFECTS = 1;
     */
    MethodOptions_IdempotencyLevel[MethodOptions_IdempotencyLevel["NO_SIDE_EFFECTS"] = 1] = "NO_SIDE_EFFECTS";
    /**
     * idempotent, but may have side effects
     *
     * @generated from protobuf enum value: IDEMPOTENT = 2;
     */
    MethodOptions_IdempotencyLevel[MethodOptions_IdempotencyLevel["IDEMPOTENT"] = 2] = "IDEMPOTENT";
})(MethodOptions_IdempotencyLevel || (MethodOptions_IdempotencyLevel = {}));
/**
 * Type for protobuf message google.protobuf.FileDescriptorSet
 */
class FileDescriptorSet$Type extends MessageType {
    constructor() {
        super("google.protobuf.FileDescriptorSet", [
            { no: 1, name: "file", kind: "message", repeat: RepeatType.UNPACKED, T: () => FileDescriptorProto }
        ]);
    }
}
export const FileDescriptorSet = new FileDescriptorSet$Type();
/**
 * Type for protobuf message google.protobuf.FileDescriptorProto
 */
class FileDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.FileDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "package", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 3, name: "dependency", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.STRING },
            { no: 10, name: "public_dependency", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.INT32 },
            { no: 11, name: "weak_dependency", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.INT32 },
            { no: 4, name: "message_type", kind: "message", repeat: RepeatType.UNPACKED, T: () => DescriptorProto },
            { no: 5, name: "enum_type", kind: "message", repeat: RepeatType.UNPACKED, T: () => EnumDescriptorProto },
            { no: 6, name: "service", kind: "message", repeat: RepeatType.UNPACKED, T: () => ServiceDescriptorProto },
            { no: 7, name: "extension", kind: "message", repeat: RepeatType.UNPACKED, T: () => FieldDescriptorProto },
            { no: 8, name: "options", kind: "message", T: () => FileOptions },
            { no: 9, name: "source_code_info", kind: "message", T: () => SourceCodeInfo },
            { no: 12, name: "syntax", kind: "scalar", opt: true, T: ScalarType.STRING }
        ]);
    }
}
export const FileDescriptorProto = new FileDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.DescriptorProto
 */
class DescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.DescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "field", kind: "message", repeat: RepeatType.UNPACKED, T: () => FieldDescriptorProto },
            { no: 6, name: "extension", kind: "message", repeat: RepeatType.UNPACKED, T: () => FieldDescriptorProto },
            { no: 3, name: "nested_type", kind: "message", repeat: RepeatType.UNPACKED, T: () => DescriptorProto },
            { no: 4, name: "enum_type", kind: "message", repeat: RepeatType.UNPACKED, T: () => EnumDescriptorProto },
            { no: 5, name: "extension_range", kind: "message", repeat: RepeatType.UNPACKED, T: () => DescriptorProto_ExtensionRange },
            { no: 8, name: "oneof_decl", kind: "message", repeat: RepeatType.UNPACKED, T: () => OneofDescriptorProto },
            { no: 7, name: "options", kind: "message", T: () => MessageOptions },
            { no: 9, name: "reserved_range", kind: "message", repeat: RepeatType.UNPACKED, T: () => DescriptorProto_ReservedRange },
            { no: 10, name: "reserved_name", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.STRING }
        ]);
    }
}
export const DescriptorProto = new DescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.DescriptorProto.ExtensionRange
 */
class DescriptorProto_ExtensionRange$Type extends MessageType {
    constructor() {
        super("google.protobuf.DescriptorProto.ExtensionRange", [
            { no: 1, name: "start", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 2, name: "end", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 3, name: "options", kind: "message", T: () => ExtensionRangeOptions }
        ]);
    }
}
export const DescriptorProto_ExtensionRange = new DescriptorProto_ExtensionRange$Type();
/**
 * Type for protobuf message google.protobuf.DescriptorProto.ReservedRange
 */
class DescriptorProto_ReservedRange$Type extends MessageType {
    constructor() {
        super("google.protobuf.DescriptorProto.ReservedRange", [
            { no: 1, name: "start", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 2, name: "end", kind: "scalar", opt: true, T: ScalarType.INT32 }
        ]);
    }
}
export const DescriptorProto_ReservedRange = new DescriptorProto_ReservedRange$Type();
/**
 * Type for protobuf message google.protobuf.ExtensionRangeOptions
 */
class ExtensionRangeOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.ExtensionRangeOptions", [
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const ExtensionRangeOptions = new ExtensionRangeOptions$Type();
/**
 * Type for protobuf message google.protobuf.FieldDescriptorProto
 */
class FieldDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.FieldDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 3, name: "number", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 4, name: "label", kind: "enum", opt: true, T: () => ["google.protobuf.FieldDescriptorProto.Label", FieldDescriptorProto_Label, "LABEL_"] },
            { no: 5, name: "type", kind: "enum", opt: true, T: () => ["google.protobuf.FieldDescriptorProto.Type", FieldDescriptorProto_Type, "TYPE_"] },
            { no: 6, name: "type_name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "extendee", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 7, name: "default_value", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 9, name: "oneof_index", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 10, name: "json_name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 8, name: "options", kind: "message", T: () => FieldOptions },
            { no: 17, name: "proto3_optional", kind: "scalar", opt: true, T: ScalarType.BOOL }
        ]);
    }
}
export const FieldDescriptorProto = new FieldDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.OneofDescriptorProto
 */
class OneofDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.OneofDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "options", kind: "message", T: () => OneofOptions }
        ]);
    }
}
export const OneofDescriptorProto = new OneofDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.EnumDescriptorProto
 */
class EnumDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.EnumDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "value", kind: "message", repeat: RepeatType.UNPACKED, T: () => EnumValueDescriptorProto },
            { no: 3, name: "options", kind: "message", T: () => EnumOptions },
            { no: 4, name: "reserved_range", kind: "message", repeat: RepeatType.UNPACKED, T: () => EnumDescriptorProto_EnumReservedRange },
            { no: 5, name: "reserved_name", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.STRING }
        ]);
    }
}
export const EnumDescriptorProto = new EnumDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.EnumDescriptorProto.EnumReservedRange
 */
class EnumDescriptorProto_EnumReservedRange$Type extends MessageType {
    constructor() {
        super("google.protobuf.EnumDescriptorProto.EnumReservedRange", [
            { no: 1, name: "start", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 2, name: "end", kind: "scalar", opt: true, T: ScalarType.INT32 }
        ]);
    }
}
export const EnumDescriptorProto_EnumReservedRange = new EnumDescriptorProto_EnumReservedRange$Type();
/**
 * Type for protobuf message google.protobuf.EnumValueDescriptorProto
 */
class EnumValueDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.EnumValueDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "number", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 3, name: "options", kind: "message", T: () => EnumValueOptions }
        ]);
    }
}
export const EnumValueDescriptorProto = new EnumValueDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.ServiceDescriptorProto
 */
class ServiceDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.ServiceDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "method", kind: "message", repeat: RepeatType.UNPACKED, T: () => MethodDescriptorProto },
            { no: 3, name: "options", kind: "message", T: () => ServiceOptions }
        ]);
    }
}
export const ServiceDescriptorProto = new ServiceDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.MethodDescriptorProto
 */
class MethodDescriptorProto$Type extends MessageType {
    constructor() {
        super("google.protobuf.MethodDescriptorProto", [
            { no: 1, name: "name", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 2, name: "input_type", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 3, name: "output_type", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 4, name: "options", kind: "message", T: () => MethodOptions },
            { no: 5, name: "client_streaming", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 6, name: "server_streaming", kind: "scalar", opt: true, T: ScalarType.BOOL }
        ]);
    }
}
export const MethodDescriptorProto = new MethodDescriptorProto$Type();
/**
 * Type for protobuf message google.protobuf.FileOptions
 */
class FileOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.FileOptions", [
            { no: 1, name: "java_package", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 8, name: "java_outer_classname", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 10, name: "java_multiple_files", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 20, name: "java_generate_equals_and_hash", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 27, name: "java_string_check_utf8", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 9, name: "optimize_for", kind: "enum", opt: true, T: () => ["google.protobuf.FileOptions.OptimizeMode", FileOptions_OptimizeMode] },
            { no: 11, name: "go_package", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 16, name: "cc_generic_services", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 17, name: "java_generic_services", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 18, name: "py_generic_services", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 42, name: "php_generic_services", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 23, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 31, name: "cc_enable_arenas", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 36, name: "objc_class_prefix", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 37, name: "csharp_namespace", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 39, name: "swift_prefix", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 40, name: "php_class_prefix", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 41, name: "php_namespace", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 44, name: "php_metadata_namespace", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 45, name: "ruby_package", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const FileOptions = new FileOptions$Type();
/**
 * Type for protobuf message google.protobuf.MessageOptions
 */
class MessageOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.MessageOptions", [
            { no: 1, name: "message_set_wire_format", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 2, name: "no_standard_descriptor_accessor", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 3, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 7, name: "map_entry", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const MessageOptions = new MessageOptions$Type();
/**
 * Type for protobuf message google.protobuf.FieldOptions
 */
class FieldOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.FieldOptions", [
            { no: 1, name: "ctype", kind: "enum", opt: true, T: () => ["google.protobuf.FieldOptions.CType", FieldOptions_CType] },
            { no: 2, name: "packed", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 6, name: "jstype", kind: "enum", opt: true, T: () => ["google.protobuf.FieldOptions.JSType", FieldOptions_JSType] },
            { no: 5, name: "lazy", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 3, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 10, name: "weak", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const FieldOptions = new FieldOptions$Type();
/**
 * Type for protobuf message google.protobuf.OneofOptions
 */
class OneofOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.OneofOptions", [
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const OneofOptions = new OneofOptions$Type();
/**
 * Type for protobuf message google.protobuf.EnumOptions
 */
class EnumOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.EnumOptions", [
            { no: 2, name: "allow_alias", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 3, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const EnumOptions = new EnumOptions$Type();
/**
 * Type for protobuf message google.protobuf.EnumValueOptions
 */
class EnumValueOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.EnumValueOptions", [
            { no: 1, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const EnumValueOptions = new EnumValueOptions$Type();
/**
 * Type for protobuf message google.protobuf.ServiceOptions
 */
class ServiceOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.ServiceOptions", [
            { no: 33, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const ServiceOptions = new ServiceOptions$Type();
/**
 * Type for protobuf message google.protobuf.MethodOptions
 */
class MethodOptions$Type extends MessageType {
    constructor() {
        super("google.protobuf.MethodOptions", [
            { no: 33, name: "deprecated", kind: "scalar", opt: true, T: ScalarType.BOOL },
            { no: 34, name: "idempotency_level", kind: "enum", opt: true, T: () => ["google.protobuf.MethodOptions.IdempotencyLevel", MethodOptions_IdempotencyLevel] },
            { no: 999, name: "uninterpreted_option", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption }
        ]);
    }
}
export const MethodOptions = new MethodOptions$Type();
/**
 * Type for protobuf message google.protobuf.UninterpretedOption
 */
class UninterpretedOption$Type extends MessageType {
    constructor() {
        super("google.protobuf.UninterpretedOption", [
            { no: 2, name: "name", kind: "message", repeat: RepeatType.UNPACKED, T: () => UninterpretedOption_NamePart },
            { no: 3, name: "identifier_value", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 4, name: "positive_int_value", kind: "scalar", opt: true, T: ScalarType.UINT64 },
            { no: 5, name: "negative_int_value", kind: "scalar", opt: true, T: ScalarType.INT64 },
            { no: 6, name: "double_value", kind: "scalar", opt: true, T: ScalarType.DOUBLE },
            { no: 7, name: "string_value", kind: "scalar", opt: true, T: ScalarType.BYTES },
            { no: 8, name: "aggregate_value", kind: "scalar", opt: true, T: ScalarType.STRING }
        ]);
    }
}
export const UninterpretedOption = new UninterpretedOption$Type();
/**
 * Type for protobuf message google.protobuf.UninterpretedOption.NamePart
 */
class UninterpretedOption_NamePart$Type extends MessageType {
    constructor() {
        super("google.protobuf.UninterpretedOption.NamePart", [
            { no: 1, name: "name_part", kind: "scalar", T: ScalarType.STRING },
            { no: 2, name: "is_extension", kind: "scalar", T: ScalarType.BOOL }
        ]);
    }
}
export const UninterpretedOption_NamePart = new UninterpretedOption_NamePart$Type();
/**
 * Type for protobuf message google.protobuf.SourceCodeInfo
 */
class SourceCodeInfo$Type extends MessageType {
    constructor() {
        super("google.protobuf.SourceCodeInfo", [
            { no: 1, name: "location", kind: "message", repeat: RepeatType.UNPACKED, T: () => SourceCodeInfo_Location }
        ]);
    }
}
export const SourceCodeInfo = new SourceCodeInfo$Type();
/**
 * Type for protobuf message google.protobuf.SourceCodeInfo.Location
 */
class SourceCodeInfo_Location$Type extends MessageType {
    constructor() {
        super("google.protobuf.SourceCodeInfo.Location", [
            { no: 1, name: "path", kind: "scalar", repeat: RepeatType.PACKED, T: ScalarType.INT32 },
            { no: 2, name: "span", kind: "scalar", repeat: RepeatType.PACKED, T: ScalarType.INT32 },
            { no: 3, name: "leading_comments", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 4, name: "trailing_comments", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 6, name: "leading_detached_comments", kind: "scalar", repeat: RepeatType.UNPACKED, T: ScalarType.STRING }
        ]);
    }
}
export const SourceCodeInfo_Location = new SourceCodeInfo_Location$Type();
/**
 * Type for protobuf message google.protobuf.GeneratedCodeInfo
 */
class GeneratedCodeInfo$Type extends MessageType {
    constructor() {
        super("google.protobuf.GeneratedCodeInfo", [
            { no: 1, name: "annotation", kind: "message", repeat: RepeatType.UNPACKED, T: () => GeneratedCodeInfo_Annotation }
        ]);
    }
}
export const GeneratedCodeInfo = new GeneratedCodeInfo$Type();
/**
 * Type for protobuf message google.protobuf.GeneratedCodeInfo.Annotation
 */
class GeneratedCodeInfo_Annotation$Type extends MessageType {
    constructor() {
        super("google.protobuf.GeneratedCodeInfo.Annotation", [
            { no: 1, name: "path", kind: "scalar", repeat: RepeatType.PACKED, T: ScalarType.INT32 },
            { no: 2, name: "source_file", kind: "scalar", opt: true, T: ScalarType.STRING },
            { no: 3, name: "begin", kind: "scalar", opt: true, T: ScalarType.INT32 },
            { no: 4, name: "end", kind: "scalar", opt: true, T: ScalarType.INT32 }
        ]);
    }
}
export const GeneratedCodeInfo_Annotation = new GeneratedCodeInfo_Annotation$Type();
